driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java

```diff
@@ -96,108 +112,151 @@ public void connect() {
-                    connection.close();
+
+                if (currentCheckCancelled) {
+                    waitForNext();
+                    currentCheckCancelled = false;
+                    continue;
                 }
+
+                logStateChange(previousServerDescription, currentServerDescription);
+                serverStateListener.stateChanged(new ChangeEvent<>(previousServerDescription, currentServerDescription));
+
+                if (currentServerDescription.getException() != null) {
+                    connectionPool.invalidate();
+                }
+
+                if (((connection == null || shouldStreamResponses(currentServerDescription))
+                        && currentServerDescription.getTopologyVersion() != null)
+                        || (connection != null && connection.hasMoreToCome())
+                        || (currentServerDescription.getException() instanceof MongoSocketException
+                        && previousServerDescription.getType() != UNKNOWN)) {
+                    continue;
+                }
+                waitForNext();
             }
         }
 
         private ServerDescription getConnectingServerDescription(final Throwable exception) {
             return ServerDescription.builder().type(UNKNOWN).state(CONNECTING).address(serverId.getAddress()).exception(exception).build();
         }
 
-        private ServerDescription lookupServerDescription(final InternalConnection connection) {
-            if (LOGGER.isDebugEnabled()) {
-                LOGGER.debug(format("Checking status of %s", serverId.getAddress()));
-            }
-            serverMonitorListener.serverHearbeatStarted(new ServerHeartbeatStartedEvent(connection.getDescription().getConnectionId()));
-
-            long start = System.nanoTime();
+        private ServerDescription lookupServerDescription(final ServerDescription currentServerDescription) {
             try {
-                BsonDocument isMasterResult =
-                        executeCommand("admin", new BsonDocument("ismaster", new BsonInt32(1)), clusterClock, connection);
-                long elapsedTimeNanos = System.nanoTime() - start;
-                averageRoundTripTime.addSample(elapsedTimeNanos);
-
-                serverMonitorListener.serverHeartbeatSucceeded(
-                        new ServerHeartbeatSucceededEvent(connection.getDescription().getConnectionId(), isMasterResult, elapsedTimeNanos));
-
-                return createServerDescription(serverId.getAddress(), isMasterResult, averageRoundTripTime.getAverage());
-            } catch (RuntimeException e) {
-                serverMonitorListener.serverHeartbeatFailed(
-                        new ServerHeartbeatFailedEvent(connection.getDescription().getConnectionId(), System.nanoTime() - start, e));
-                throw e;
+                if (connection == null || connection.isClosed()) {
+                    currentCheckCancelled = false;
```

Comment:
After talking with Divjot, decided that this variable should be reset here in order to avoid unnecessarily going into the wait state twice in a row after a cancel.

Reply:
Good catch